x-pack/plugins/ml/public/components/anomalies_table/anomalies_table.js

```diff
@@ -0,0 +1,402 @@
+      const matching = nextProps.tableData.anomalies.find((anomaly) => {
+        return anomaly.rowId === rowId;
+      });
+
+      return matching === undefined;
+    });
+
+    if (prevExpandedNotInData !== undefined) {
+      // Anomaly data has changed and an anomaly that was previously expanded is no longer in the data.
+      return {
+        itemIdToExpandedRowMap: {}
+      };
+    }
+
+    // Return null to indicate no change to state.
+    return null;
+  }
+
+  toggleRow = (item) => {
+    const itemIdToExpandedRowMap = { ...this.state.itemIdToExpandedRowMap };
+    if (itemIdToExpandedRowMap[item.rowId]) {
+      delete itemIdToExpandedRowMap[item.rowId];
+    } else {
+      const examples = item.entityName === 'mlcategory' ?
+        _.get(this.props.tableData, ['examplesByJobId', item.jobId, item.entityValue]) : undefined;
+      itemIdToExpandedRowMap[item.rowId] = (
+        <AnomalyDetails
+          anomaly={item}
+          examples={examples}
+          isAggregatedData={this.isShowingAggregatedData()}
+          filter={this.props.filter}
+        />
+      );
+    }
+    this.setState({ itemIdToExpandedRowMap });
+  };
+
+  onMouseOver = (event) => {
+    // Triggered when the mouse is somewhere over the table.
+    // Traverse through the table DOM to find the expand/collapse
+    // button which stores the ID of the row.
+    let mouseOverRecord = undefined;
+    const target = $(event.target);
+    const parentRow = target.closest('tr');
+    const firstCell = parentRow.children('td').first();
+    if (firstCell !== undefined) {
+      const expandButton = firstCell.find('button').first();
+      if (expandButton.length > 0) {
+        const rowId = expandButton.attr('data-row-id');
+        mouseOverRecord = this.props.tableData.anomalies.find((anomaly) => {
+          return anomaly.rowId === rowId;
+        });
+      }
+    }
+
+    if (this.mouseOverRecord !== undefined) {
+      if (mouseOverRecord === undefined || this.mouseOverRecord.rowId !== mouseOverRecord.rowId) {
+        // Mouse is over a different row, fire mouseleave on the previous record.
+        mlAnomaliesTableService.anomalyRecordMouseleave.changed(this.mouseOverRecord);
+
+        if (mouseOverRecord !== undefined) {
+          // Mouse is over a new row, fire mouseenter on the new record.
+          mlAnomaliesTableService.anomalyRecordMouseenter.changed(mouseOverRecord);
+        }
+      }
+    } else {
+      if (mouseOverRecord !== undefined) {
+        // Mouse is now over a row, fire mouseenter on the record.
+        mlAnomaliesTableService.anomalyRecordMouseenter.changed(mouseOverRecord);
+      }
+    }
+
+    this.mouseOverRecord = mouseOverRecord;
+  };
+
+  onMouseLeave = () => {
+    if (this.mouseOverRecord !== undefined) {
+      mlAnomaliesTableService.anomalyRecordMouseleave.changed(this.mouseOverRecord);
+    }
+  };
+
+  render() {
+    if (this.props.tableData === undefined ||
```

Comment:
`const { timefilter, tableData, filter } = this.props;` could be added to the beginning of the `render()` method so in the code below `this.props` wouldn't be necessary all the time. This would make it clearer which props are used in the `render()` method. See section 4 in this article with more details about this pattern https://medium.com/walmartlabs/make-your-react-components-pretty-a1ae4ec0f56e

Reply:
I like this pattern. as well as making it clear what props are being used in the function, it also cleans up the code throughout the function.